web/pdf_history.js

```diff
@@ -13,21 +13,455 @@
+    this._popStateInProgress = true;
+
+    if (hashChanged) {
+      // When the hash changed, implying that the 'popstate' event will be
+      // followed by a 'hashchange' event, then we do *not* want to update the
+      // browser history when handling the 'hashchange' event (in web/app.js)
+      // since that would *overwrite* the new destination navigated to below.
+      //
+      // To avoid accidentally disabling all future user-initiated hash changes,
+      // if there's e.g. another 'hashchange' listeners that stops the event
+      // propagation, we make sure to always force-reset `this._blockHashChange`
+      // after `HASH_CHANGE_TIMEOUT` milliseconds has passed.
+      this._blockHashChange++;
+      waitOnEventOrTimeout({
+        target: window,
+        name: 'hashchange',
+        delay: HASH_CHANGE_TIMEOUT,
+      }).then((type) => {
+        this._blockHashChange--;
+      });
+    }
 
-  back() {}
+    // This case corresponds to navigation backwards in the browser history.
+    if (state.uid < this._currentUid && this._position && this._destination) {
+      if (this._destination.page &&
+          this._destination.page !== this._position.first &&
+          this._destination.page !== this._position.page) {
+        // If the `page` of the `this._destination` is no longer visible,
+        // push the `this._position` to the browser history before moving back.
+        this._pushOrReplaceState(this._destination);
+        this._pushOrReplaceState(this._position);
 
-  forward() {}
+        // After `window.history.back()`, we must not enter this block on the
+        // resulting 'popstate' event, since that may cause an infinite loop.
+        this._currentUid = state.uid;
+
+        window.history.back();
+        return;
+      }
+    }
+
+    // Navigate to the new destination.
+    let destination = state.destination;
+    this._resetSomeInternalState(destination, state.uid);
+
+    if (destination.dest) {
+      this.linkService.navigateTo(destination.dest);
+    } else if (destination.hash) {
+      this.linkService.setHash(destination.hash);
+    } else if (destination.page) {
+      // Fallback case; shouldn't be necessary, but better safe than sorry.
+      this.linkService.page = destination.page;
+    }
+
+    // Since `PDFLinkService.navigateTo` is asynchronous, we thus defer the
+    // resetting of `this._popStateInProgress` slightly.
+    Promise.resolve().then(() => {
+      this._popStateInProgress = false;
+    });
+  }
+
+  /**
+   * @private
+   */
+  _bindEvents() {
+    let { _boundEvents, eventBus, } = this;
+
+    _boundEvents.updateViewarea = this._updateViewarea.bind(this);
+    _boundEvents.popState = this._popState.bind(this);
+    _boundEvents.pageHide = (evt) => {
+      // Attempt to push the `this._position` into the browser history when
+      // navigating away from the document, if the history is currently empty.
```

Comment:
Why only if the history is empty? If there are history items, I change the position and navigate away from the document, I would expect the last position to be retained when I navigate back.

Reply:
By the time the `pagehide` event fires, it's too late to make (most) modifications of the history. That is also the whole reason for the existence of commit https://github.com/mozilla/pdf.js/pull/8775/commits/281070b20c3ed17979b85482922d9245c8eee623, refer to the big comment starting at https://github.com/mozilla/pdf.js/pull/8775/commits/281070b20c3ed17979b85482922d9245c8eee623#diff-3b4ebbe223f93c9123a754b4845844e7R380.

At this point in time, since the viewer is already being unloaded, it's impossible to push a *new* entry to the browser history. What would happen instead here, is that the current browser history entry would be overwritten (which is definitely *not* what you want here).
The only case where that doesn't matter is if the history is empty, and for all other cases we'll now rely on the temporary position instead here.

I'll try to amend this comment a bit, to make this clearer.